Eu errei muito nesse, fui querer pagar de bom e fiz o header = None e fudeu tudo e transformou todas as colunas em Obj o que dificultou muito

In [1]:
#Aqui vamos resolver o trabalho prático do módulo 2 
import pandas as pd
import numpy as np

In [2]:
#Agora que importamos as bibliotecas, vamos upar nosso CSV
from google.colab import files
upload_bike = files.upload()

Saving bike-sharing.csv to bike-sharing.csv


In [164]:
#Agora vamos fazer a leitura desse CSV
dataset_bike =  pd.read_csv("https://pycourse.s3.amazonaws.com/bike-sharing.csv", header = None)
dataset_bike.head(20)

0       1       2   ...      14          15           16
0     datetime  rec_id  season  ...  casual  registered  total_count
1   2011-01-01       1       1  ...       3          13           16
2   2011-01-01       2       1  ...       8          32           40
3   2011-01-01       3       1  ...       5          27           32
4   2011-01-01       4       1  ...       3          10           13
5   2011-01-01       5       1  ...       0           1            1
6   2011-01-01       6       1  ...       0           1            1
7   2011-01-01       7       1  ...       2           0            2
8   2011-01-01       8       1  ...       1           2            3
9   2011-01-01       9       1  ...       1           7            8
10  2011-01-01      10       1  ...       8           6           14
11  2011-01-01      11       1  ...      12          24           36
12  2011-01-01      12       1  ...      26          30           56
13  2011-01-01      13       1  ...      29          55           84
14  2011-01-01      14       1  ...      47          47           94
15  2011-01-01      15       1  ...      35          71          106
16  2011-01-01      16       1  ...      40          70          110
17  2011-01-01      17       1  ...      41          52           93
18  2011-01-01      18       1  ...      15          52           67
19  2011-01-01      19       1  ...       9          26           35

[20 rows x 17 columns]

0. datetime: data;
1. rec_id: índice do registro de locação;
2. season: estação do ano (1: inverno, 2: primavera, 3: verão, 4: outono).     Relativo aohemisfério norte;
3. year: ano (0: 2011, 1:2012);
4. month: mês (1 a 12);
5. hour: hora do dia (0 a 23);
6. is_holiday: booleano indicando feriado;
7. weekday: dia da semana (0: domingo, 1: segunda-feira, …, 6: sábado);
8. is_workingday: booleano indicando dia útil;
9. weather_condition: (1: limpo, 2: nublado, 3: chuva leve, 4: chuva forte);
10. temp: temperatura escalada entre 0 e 1. Valor original em graus Celsius: -8 a 39;
11. atemp: sensação térmica escalada entre 0 e 1. Valor original em graus Celsius: -16 a 50;
12. humidity: humidade relativa (0 a 1);
13. windspeed: velocidade do vento escalada entre 0 e 1 (máximo original: 67);
14. casual: número de locações para usuários casuais;
15. registered: número de locações para usuários registrados;
16. total_count: contador total de aluguéis (casual+registered).


Agora que sabemos o que cada colunas nos diz, vamos começar a responder as perguntas: a. Qual o tamanho desse dataset? <br> R: (17380, 17)

In [165]:
#Vamos descobrir o tamanho do dataset com .infor()
dataset_bike.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17380 entries, 0 to 17379
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       17380 non-null  object
 1   1       17380 non-null  object
 2   2       17380 non-null  object
 3   3       17380 non-null  object
 4   4       17380 non-null  object
 5   5       17380 non-null  object
 6   6       17380 non-null  object
 7   7       17380 non-null  object
 8   8       17380 non-null  object
 9   9       17380 non-null  object
 10  10      17380 non-null  object
 11  11      17380 non-null  object
 12  12      17380 non-null  object
 13  13      17380 non-null  object
 14  14      17380 non-null  object
 15  15      17380 non-null  object
 16  16      17380 non-null  object
dtypes: object(17)
memory usage: 2.3+ MB


b. Qual a média da coluna windspeed?R: 0,49698<br>
c. Qual a média da coluna temp?R:0,1900

In [166]:
#Para conseguir informções de dados estatísticos vamos usar o describe()
dataset_bike.describe()

0      1      2      3   ...     13     14     15     16
count        17380  17380  17380  17380  ...  17380  17380  17380  17380
unique         732  17380      5      3  ...     31    323    777    870
top     2012-09-10  10015      3      1  ...    0.0      0      4      5
freq            24      1   4496   8734  ...   2180   1581    307    260

[4 rows x 17 columns]

In [167]:
#Como não temos informações a partir do describe, vamos mudar a forma do dado
#Vamos transformar os dados de forma que possamos analisa-los
dataset_bike_drop = dataset_bike.drop([0], axis = 0, errors='ignore')
dataset_bike_drop.head()

0  1  2  3  4  5  6  7  8  9     10      11    12   13 14  15  16
1  2011-01-01  1  1  0  1  0  0  6  0  1  0.24  0.2879  0.81  0.0  3  13  16
2  2011-01-01  2  1  0  1  1  0  6  0  1  0.22  0.2727   0.8  0.0  8  32  40
3  2011-01-01  3  1  0  1  2  0  6  0  1  0.22  0.2727   0.8  0.0  5  27  32
4  2011-01-01  4  1  0  1  3  0  6  0  1  0.24  0.2879  0.75  0.0  3  10  13
5  2011-01-01  5  1  0  1  4  0  6  0  1  0.24  0.2879  0.75  0.0  0   1   1

In [114]:
dataset_bike_drop[10] = pd.to_numeric(dataset_bike_drop[10]) #transformando os valores das colunas em float
dataset_bike_drop[13] = pd.to_numeric(dataset_bike_drop[13])
dataset_bike_drop.head()

0  1  2  3  4  5  6  7  8  9     10      11    12   13 14  15  16
1  2011-01-01  1  1  0  1  0  0  6  0  1  0.24  0.2879  0.81  0.0  3  13  16
2  2011-01-01  2  1  0  1  1  0  6  0  1  0.22  0.2727   0.8  0.0  8  32  40
3  2011-01-01  3  1  0  1  2  0  6  0  1  0.22  0.2727   0.8  0.0  5  27  32
4  2011-01-01  4  1  0  1  3  0  6  0  1  0.24  0.2879  0.75  0.0  3  10  13
5  2011-01-01  5  1  0  1  4  0  6  0  1  0.24  0.2879  0.75  0.0  0   1   1

In [115]:
dataset_bike_drop.info() #verificando informações do dataset
dataset_bike_drop.describe()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17379 entries, 1 to 17379
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       17379 non-null  object 
 1   1       17379 non-null  object 
 2   2       17379 non-null  object 
 3   3       17379 non-null  object 
 4   4       17379 non-null  object 
 5   5       17379 non-null  object 
 6   6       17379 non-null  object 
 7   7       17379 non-null  object 
 8   8       17379 non-null  object 
 9   9       17379 non-null  object 
 10  10      17379 non-null  float64
 11  11      17379 non-null  object 
 12  12      17379 non-null  object 
 13  13      17379 non-null  float64
 14  14      17379 non-null  object 
 15  15      17379 non-null  object 
 16  16      17379 non-null  object 
dtypes: float64(2), object(15)
memory usage: 2.4+ MB


10            13
count  17379.000000  17379.000000
mean       0.496987      0.190098
std        0.192556      0.122340
min        0.020000      0.000000
25%        0.340000      0.104500
50%        0.500000      0.194000
75%        0.660000      0.253700
max        1.000000      0.850700

d. Quantos registros existem para o ano de 2011?R:8645 <br>
e. Quantos registros existem para o ano de 2012?R:8734

In [116]:
#Sabemos que os dados foram categorizados e deixados como objetos e que 0:2011 e 1:2012
#Vamos transformar a coluna em um vetor de numeros e fazer a contagem
dataset_bike_drop[3].astype('str').value_counts()

1    8734
0    8645
Name: 3, dtype: int64

f. Quantas locações de bicicletas foram efetuadas em 2011?R:123103<br>
g. Quantas locações de bicicletas foram efetuadas em 2012?R:2049576

In [117]:
#Para poder resolver esse problema, vamos selecionar a condição e a coluna e somar
print(dataset_bike_drop[dataset_bike_drop[3] == '0'][16].astype(int).sum())
print(dataset_bike_drop[dataset_bike_drop[3] == '1'][16].astype(int).sum())

1243103
2049576


h. Qual estação do ano contém a maior média de locações de bicicletas?R: 3 - Verão<br>
i. Qual estação do ano contém a menor média de locações de bicicletas?R: 1 - Inverno

In [118]:
#Agora, vamos selecionar a condição que queremos e criar 3 vetores para cada estação e verificar a maior média
#Primeiro, vamos verificar os valores unicos
dataset_bike_drop[2].unique()

array(['1', '2', '3', '4'], dtype=object)

temos:
1. inverno
2. Primavera
3. Verão
4. Outono

In [119]:
#Criando vetores e printando a média
print(dataset_bike_drop[dataset_bike_drop[2] == '1'][16].astype(int).mean())
print(dataset_bike_drop[dataset_bike_drop[2] == '2'][16].astype(int).mean())
print(dataset_bike_drop[dataset_bike_drop[2] == '3'][16].astype(int).mean())
print(dataset_bike_drop[dataset_bike_drop[2] == '4'][16].astype(int).mean())

111.11456859971712
208.34406894987526
236.01623665480426
198.86885633270322


j. Qual horário do dia contém a maior média de locações de bicicletas?R:17 <br>
k. Qual horário do dia contém a menor média de locações de bicicletas?R:4

In [128]:
#Para resolvermos esse problema, vamos construir um vetor com a média de todos os horários e em seguida verificar o maior e o menor
for i in range(0, 24, 1):
  print(i)
  print(dataset_bike_drop[dataset_bike_drop[5].astype(int) == i][16].astype(int).mean())

0
53.89807162534435
1
33.3756906077348
2
22.86993006993007
3
11.727403156384504
4
6.352941176470588
5
19.88981868898187
6
76.04413793103448
7
212.0646492434663
8
359.01100412654745
9
219.30949105914718
10
173.6685006877579
11
208.1430536451169
12
253.31593406593407
13
253.66117969821673
14
240.94924554183814
15
251.2331961591221
16
311.9835616438356
17
461.45205479452056
18
425.510989010989
19
311.52335164835165
20
226.03021978021977
21
172.31456043956044
22
131.33516483516485
23
87.83104395604396


l. Que dia da semana contém a maior média de locações de bicicletas?R: 4 - Quinta <br>
m. Que dia da semana contém a menor média de locações de bicicletas?R:0 - Domingo 


In [148]:
#Para resolver esse problema, vamos fazer o mesmo do caso anterior
#weekday: dia da semana (0: domingo, 1: segunda-feira, …, 6: sábado);
for i in range(0, 7, 1):
  print(i)
  print(dataset_bike_drop[dataset_bike_drop[5].astype(int) == i][16].astype(int).mean())

0
53.89807162534435
1
33.3756906077348
2
22.86993006993007
3
11.727403156384504
4
6.352941176470588
5
19.88981868898187
6
76.04413793103448


n. Às quartas-feiras (weekday = 3), qual o horário do dia contém a maior média
de locações de bicicletas?<br>
o. Aos sábados (weekday = 6), qual o horário do dia contém a maior média de
locações de bicicletas?

In [172]:
#Agora que o bixo pega
dataset_bike_drop[7] = dataset_bike_drop[7].astype(int)
dataset_bike_drop[5] = dataset_bike_drop[5].astype(int)
dataset_bike_drop[16] = dataset_bike_drop[16].astype(int)
dataset_bike_drop.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17379 entries, 1 to 17379
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       17379 non-null  object
 1   1       17379 non-null  object
 2   2       17379 non-null  object
 3   3       17379 non-null  object
 4   4       17379 non-null  object
 5   5       17379 non-null  int64 
 6   6       17379 non-null  object
 7   7       17379 non-null  int64 
 8   8       17379 non-null  object
 9   9       17379 non-null  object
 10  10      17379 non-null  object
 11  11      17379 non-null  object
 12  12      17379 non-null  object
 13  13      17379 non-null  object
 14  14      17379 non-null  object
 15  15      17379 non-null  object
 16  16      17379 non-null  int64 
dtypes: int64(3), object(14)
memory usage: 2.4+ MB


In [179]:
print(dataset_bike_drop[dataset_bike_drop[7] == 3].groupby(5).agg('mean')[16].sort_values(ascending=False).head(1))
print(dataset_bike_drop[dataset_bike_drop[7] == 6].groupby(5).agg('mean')[16].sort_values(ascending=False).head(1))

5
17    513.144231
Name: 16, dtype: float64
5
13    385.371429
Name: 16, dtype: float64
